ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -253,56 +253,62 @@ public boolean updateAccounts(Collection<Account> accounts) {
    * {@inheritDoc}
    */
   @Override
-  public Pair<String, String> addContainer(String accountName, Container newContainer) {
+  public Collection<Container> addContainers(String accountName, Collection<Container> containers) {
     checkOpen();
-    if (accountName == null || accountName.isEmpty() || newContainer == null || newContainer.getName() == null
-        || newContainer.getName().isEmpty()) {
-      logger.error("Account name: {} or container: {} is invalid", accountName, newContainer);
-      throw new IllegalArgumentException("Account or container name is null or empty");
+    // input validation
+    if (accountName == null || accountName.isEmpty() || containers == null || containers.isEmpty()) {
+      throw new IllegalArgumentException("Account or container is null or empty");
     }
     Account account = getAccountByName(accountName);
     if (account == null) {
       logger.error("Account {} is not found", accountName);
       throw new IllegalArgumentException("Account " + accountName + " is not found");
     }
-    String accountId = String.valueOf(account.getId());
-    for (Container container : account.getAllContainers()) {
-      // make sure there is no conflict container (conflict means a container with same name but different attributes already exists).
-      if (container.getName().equals(newContainer.getName())) {
-        if (container.isSameContainer(newContainer)) {
-          // If an exactly same container already exists, we directly return its id. Adding same container multiple times
-          // should be no-op.
-          return new Pair<>(accountId, String.valueOf(container.getId()));
-        } else {
-          logger.error("A container: {} already exists with different attributes", container.getName());
-          throw new IllegalArgumentException("There is a conflicting container in account " + accountName);
+    Set<Container> existingContainers = new HashSet<>();
+    Set<Container> newContainers = new HashSet<>(containers);
+    for (Container existingContainer : account.getAllContainers()) {
+      for (Container newContainer : containers) {
```

Comment:
May be we can make it O(n) by storing existing container names in hash set?

Reply:
Good point, thanks for the suggestion. Will make changes based on hashmap.